pkg/ipamd/datastore/data_store.go

```diff
@@ -655,67 +563,59 @@ func (ds *DataStore) AssignPodIPv4Address(ipamKey IPAMKey) (ipv4address string,
 
 	ds.log.Debugf("AssignIPv4Address: IP address pool stats: total: %d, assigned %d", ds.total, ds.assigned)
 
-	if eni, addr := ds.eniPool.FindAddressForSandbox(ipamKey); addr != nil {
+	if eni, _, addr := ds.eniPool.FindAddressForSandbox(ipamKey); addr != nil {
 		ds.log.Infof("AssignPodIPv4Address: duplicate pod assign for sandbox %s", ipamKey)
 		return addr.Address, eni.DeviceNumber, nil
 	}
 
 	for _, eni := range ds.eniPool {
-		ds.log.Debugf("Checking for eni %v", eni)
-		if !eni.IsPDEnabled {
-			for _, addr := range eni.IPv4Addresses {
-				if !addr.Assigned() && !addr.inCoolingPeriod() {
-					ds.assignPodIPv4AddressUnsafe(ipamKey, eni, addr)
-					if err := ds.writeBackingStoreUnsafe(); err != nil {
-						ds.log.Warnf("Failed to update backing store: %v", err)
-						// Important! Unwind assignment
-						ds.unassignPodIPv4AddressUnsafe(addr)
-						return "", -1, err
-					}
-
-					return addr.Address, eni.DeviceNumber, nil
+		for _, availableCidr := range eni.AvailableIPv4Cidrs {
+			var addr *AddressInfo
+			var strPrivateIPv4 string
+			var err error
+
+			if ds.isPDEnabled && availableCidr.IsPrefix {
+				strPrivateIPv4, err = ds.getFreeIPv4AddrfromCidr(availableCidr)
+				if err != nil {
+					ds.log.Debugf("Unable to get IP address from prefix: %v", err)
+					//Check in next CIDR
+					continue
+				}
+				ds.log.Debugf("New IP from PD pool- %s", strPrivateIPv4)
+				if availableCidr.IPv4Addresses == nil {
+					availableCidr.IPv4Addresses = make(map[string]*AddressInfo)
 				}
+			} else if !ds.isPDEnabled && !availableCidr.IsPrefix {
+				strPrivateIPv4 = availableCidr.Cidr.IP.String()
+			} else {
+				//This can happen during upgrade or PD enable/disable knob toggle
+				//ENI can have prefixes attached and no space for SIPs or vice versa
+				continue
 			}
-		} else {
-			ds.log.Debugf("PD is enabled.")
-			for _, prefix := range eni.IPv4Prefixes {
-				ds.log.Debugf("Found a prefix %s", prefix)
-				if prefix.FreeIps > 0 {
-					strPrivateIPv4, err := getFreeIPv4AddrfromPrefix(prefix)
-					if err != nil {
-						ds.log.Errorf("Unable to get IP address from prefix: %v", err)
-						return "", -1, err
-					}
-					ds.log.Debugf("New IP - %s", strPrivateIPv4)
-					// Try to add the IP
-
-					err = ds.AddPrefixIPv4AddressToStore(eni.ID, strPrivateIPv4)
-					if err != nil {
-						ds.log.Errorf("Failed to ADD PD IP %s on ENI %s", strPrivateIPv4, eni)
-						return "", -1, err
-					}
-
-					//Adding here to eni DB, need to be removed while deleteing.
-					addr := eni.IPv4Addresses[strPrivateIPv4]
-					ds.assignPodIPv4AddressUnsafe(ipamKey, eni, addr)
-
-					addr.Prefix = prefix.Prefix.IP
+			addr = availableCidr.IPv4Addresses[strPrivateIPv4]
+			if addr == nil {
+				// addr is nil when we are using a new IP from prefix or SIP pool
```

Comment:
this is hacky, as it couples the implementation detail of low level data structure in high level business code.


How about a method in CIDR like, so it hides the memory handling behind CIDR structure.
```
func (i *CIDR) tryAllocateFreeIPAddress() (*AddressInfo, bool) {
     if i.IsPrefix {
        return i.tryAllocateFreeIPAddressForPrefix()
     }
    tryAllocateFreeIPAddressForSIP()
}

func (i *CIDR) tryAllocateFreeIPAddressForPrefix() (*AddressInfo, bool) {
     ...
     do the allocation
}

func (i *CIDR) tryAllocateFreeIPAddressForSIP() (*AddressInfo, bool) {
    ....
   do the allocation
}
```

Reply:
Will push this logic into tryAllicateFreeIPAddress*